src/hotspot/share/opto/phaseX.cpp

```diff
@@ -1984,7 +1985,17 @@ void PhaseCCP::push_and(Unique_Node_List& worklist, const Node* parent, const No
         push_if_not_bottom_type(worklist, n);
       }
     };
-    ConstraintCastNode::visit_uncasted_uses(use, push_and_uses_to_worklist);
+    auto push_and_uses_to_worklist_skip_conv = [&](Node* n){
+      uint opc = n->Opcode();
+      if (opc == Op_ConvI2L) {
+        // Step over a single ConvI2L in search for And. Inverse of the logic
+        // in MulNode::AndIL_shift_and_mask_is_always_zero.
+        ConstraintCastNode::visit_uncasted_uses(n, push_and_uses_to_worklist);
+      } else {
+        push_and_uses_to_worklist(n);
+      }
+    };
+    ConstraintCastNode::visit_uncasted_uses(use, push_and_uses_to_worklist_skip_conv);
```

Comment:
Maybe it would be nicer to generalize `visit_uncasted_uses`, and allow it to also peek through `ConvI2L` (optionally enabled with flag, false by default). Something like this:

```
        if (internal_use->is_ConstraintCast()) {
          internals.push(internal_use); // traverse this cast also
        } else if (peek_thru_convi2l && internal_use->Opcode() == Op_ConvI2L) {
          internals.push(internal_use); // traverse this cast also
        } else {
          callback(internal_use);
        }
```

Maybe you'd have to rename `visit_uncasted_uses` a bit. Or just create a common utility method and then call that one from `visit_uncasted_uses` with `peek_thru_convi2l = false`.

Maybe you have a better idea. But I have a sense that this may have to be done again in the future.

Reply:
I have now pushed a new version along what you suggest. A technicality is that the new version may also step over _more_ than one `ConvI2L`, which doesn't match the logic in `MulNode::AndIL_shift_and_mask_is_always_zero` that only steps over a single `ConvI2L`. However, chaining `ConvI2L` in this way does not make sense type-wise and should be caught by a type checker in some earlier stage?

Reply:
Yes, this should not be possible to go through multiple `ConvI2L`, without any `ConvL2I`. I think there would be some issues / type mismatch somewhere.